### ISSUE TITLE
Assertion failure with unary operator in subquery

### ISSUE THREAD

**Issue by hlinnaka** [greenplum-db/gpdb#1480]
```
create table foo (a integer, b integer, c text);
create table bar (a integer, b integer, c text);

create function ispositive(integer) returns bool as $$ begin return $1 > 0; end; $$ language plpgsql immutable;
create operator |>(procedure=ispositive, rightarg = integer);

postgres=# explain select * from foo where not exists (select * from bar where |>foo.a);
FATAL:  Unexpected internal error (cdbsubselect.c:271)
DETAIL:  FailedAssertion("!(list_length(opexp->args) > 1)", File: "cdbsubselect.c", Line: 271)
HINT:  Process 21065 will wait for gp_debug_linger=120 seconds before termination.
Note that its locks and other resources will not be released until then.
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
The connection to the server was lost. Attempting reset: Succeeded.
```

Spotted this while reviewing PR #1477.

**Comment by vraghavan78** [comment]
@hlinnaka the issue seem to have been fixed on the latest master.

```
vraghavan=# create table foo (a integer, b integer, c text);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
CREATE TABLE
vraghavan=# create table bar (a integer, b integer, c text);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
CREATE TABLE
vraghavan=#
vraghavan=# create function ispositive(integer) returns bool as $$ begin return $1 > 0; end; $$ language plpgsql immutable;
CREATE FUNCTION
vraghavan=# create operator |>(procedure=ispositive, rightarg = integer);
CREATE OPERATOR
vraghavan=#
vraghavan=# explain select * from foo where not exists (select * from bar where |>foo.a);
                                                      QUERY PLAN
----------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..12809.00 rows=23350 width=40)
   ->  Seq Scan on foo  (cost=0.00..12809.00 rows=7784 width=40)
         Filter: NOT ((subplan))
         SubPlan 1
           ->  Result  (cost=0.25..567.25 rows=15567 width=40)
                 One-Time Filter: |> $0
                 ->  Result  (cost=613.95..1080.95 rows=15567 width=40)
                       ->  Materialize  (cost=613.95..1080.95 rows=15567 width=40)
                             ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.25..567.25 rows=15567 width=40)
                                   ->  Seq Scan on bar  (cost=0.25..567.25 rows=15567 width=40)
 Optimizer status: legacy query optimizer
(11 rows)

vraghavan=# select * from foo where not exists (select * from bar where |>foo.a);
 a | b | c
---+---+---
(0 rows)
```

Please check and let me know if this is still an issue. If not please close the issue.

**Comment by hlinnaka** [comment]
On 01/12/2017 11:04 AM, Venkatesh Raghavan wrote:
> @hlinnaka the issue seem to have been fixed on the latest master.

I can still reproduce it on master. Are you sure you have assertions 
enabled?

- Heikki

**Comment by vraghavan78** [comment]
Sorry my bad.

**Comment by hardikar** [comment]
I think it is fixed in master:
```
shardikar=# explain select * from foo where not exists (select * from bar where |>foo.a);
                                               QUERY PLAN
--------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..853313456.96 rows=16 width=40)
   ->  Nested Loop Anti Join  (cost=0.00..853313456.96 rows=6 width=40)
         Join Filter: |> foo.a
         ->  Seq Scan on foo  (cost=0.00..567.00 rows=15567 width=40)
         ->  Materialize  (cost=0.00..3135.50 rows=46700 width=0)
               ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.00..2435.00 rows=46700 width=0)
                     ->  Seq Scan on bar  (cost=0.00..567.00 rows=15567 width=0)
 Optimizer: legacy query optimizer
(8 rows)

shardikar=# select version();
                                                                                                                       version
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 PostgreSQL 9.0beta3 (Greenplum Database 6.0.0-alpha.0+dev.5106.g43bfab4352 build dev-oss) on x86_64-apple-darwin16.7.0, compiled by GCC Apple LLVM version 9.0.0 (clang-900.0.39.2), 64-bit compiled on Apr  6 2018 10:49:08 (with assert checking)
(1 row)
```

**Comment by hlinnaka** [comment]
Yeah, this was fixed on master by commit fe2eb2c9ed3789e0dc07386202d2c8932bd54ca8. I'll leave this issue open since it still exist on GPDB 5.

**Comment by Eulerizeit** [comment]
@khannaekta since your name is on the commit that fixed this in Master ( fe2eb2c9ed ) could you take a quick look at the risk to backport the fix please?

**Comment by ivannovick** [comment]
I am closing out the github issue, but if you want to submit a patch for 5.x that is fine.
